### PR TITLE
[core] Add `requireOptionalNativeModule`

### PR DESCRIPTION
**Pull request by tsapeta**
# Why

There are already two places where we try-catch `requireNativeModule`. In doesn't look very well, so instead I propose a new `requireOptionalNativeModule` that returns `null` instead of throwing an error when the module cannot be found.

# How

- Implemented that function with a fallback to `null`
- Updated `expo-constants` and `expo-dev-menu` to use this dedicated solution instead of try-catching

# Test Plan

Tested with `expo-constants` and `expo-dev-menu`

### CODE REVIEW

File: packages/expo-modules-core/src/requireNativeModule.ts [hunk 1]
```diff
@@ -43,3 +43,16 @@ export function requireNativeModule<ModuleType = any>(moduleName: string): Modul
...
+ * Imports the native module registered with the given name. The same as `requireNativeModule`,
+ * but returns `null` when the module cannot be found instead of throwing an error.
+ *
+ * @param moduleName Name of the requested native module.
+ * @returns Object representing the native module or `null` when it cannot be found.
+ */
+export function requireOptionalNativeModule<ModuleType = any>(
+  moduleName: string
+): ModuleType | null {
+  return globalThis.expo?.modules?.[moduleName] ?? NativeModulesProxy[moduleName] ?? null;
```
Comment (Kudo) [inline packages/expo-modules-core/src/requireNativeModule.ts:57]: would be good to rewrite `requireNativeModule` to use this function and throw when the result is null. that would benefit from maintenance. e.g. when we deprecate `NativeModulesProxy`, we don't need to change two places.

**Review by Kudo — APPROVED** [review]

**Comment by meabed** [comment]
I updated to this version but i got this error 

```
 ERROR  TypeError: 0, _expoModulesCore.requireOptionalNativeModule is not a function (it is undefined), js engine: hermes

```

Downgrading to 15.0.0 fixes it.

```
{
  "dependencies": {
    "expo": "49.0.11",
    "expo-application": "5.4.0",
    "expo-asset": "8.12.0",
    "expo-background-fetch": "11.5.0",
    "expo-build-properties": "0.8.3",
    "expo-camera": "13.6.0",
    "expo-constants": "15.0.0",
    "expo-crypto": "12.6.0",
    "expo-device": "5.6.0",
    "expo-file-system": "15.6.0",
    "expo-font": "11.6.0",
    "expo-haptics": "12.6.0",
    "expo-intent-launcher": "10.7.0",
    "expo-keep-awake": "12.5.0",
    "expo-linear-gradient": "12.5.0",
    "expo-linking": "6.0.0",
    "expo-local-authentication": "13.6.0",
    "expo-localization": "14.5.0",
    "expo-location": "16.3.0",
    "expo-network": "5.6.0",
    "expo-notifications": "0.23.0",
    "expo-sharing": "11.7.0",
    "expo-splash-screen": "0.22.0",
    "expo-status-bar": "1.7.1",
    "expo-system-ui": "2.6.0",
    "expo-task-manager": "11.5.0",
    "expo-web-browser": "12.5.0"
   }
}
````

**Comment by tsapeta** [comment]
> I updated to this version

Which version exactly?

> Downgrading to 15.0.0 fixes it

Which package have you downgraded?

I'm not quite sure why do you need this feature? You can already call `requireNativeModule` in the try-catch block to achieve the same result.

**Comment by meabed** [comment]
@tsapeta 

The issue in expo-constants 15.1.0 

Using version 15.0.0 works fine.

I'm not calling the method I think it's being used there.

**Comment by meabed** [comment]
Same error for 

- expo-assets 8.12.1 https://github.com/expo/expo/blob/main/packages/expo-asset/CHANGELOG.md#8121--2023-09-16

- expo-constants 15.1.0  https://github.com/expo/expo/blob/main/packages/expo-constants/CHANGELOG.md#1510--2023-09-15

**Comment by tsapeta** [comment]
Why do you install these versions? Both are not marked as `latest` and `npx expo install` doesn't use those versions too. They are intended for SDK 50, not 49.

**Pull request merged**